### DISCUSSION TITLE
Server-client hydration with Next.js

### DISCUSSION THREAD

**Discussion by cam-eo** [reduxjs/redux-toolkit#2650] [General]
Is there a plan to cover Server-Client Hydration in RTK out of the box instead of using 3rd party packages like `next-redux-wrapper` which I see is actually mentioned [here](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering#server-side-rendering-with-nextjs). 

I'm just curious.

Because tools like [TanStack Query](https://tanstack.com/query/v4) cover this out of the box and works well but you end up needing an additional state manager for Client Side state management.

Considering the maturity of RTK I was just curious whether there is a plan to cover Server-Client hydration out the box without the need for `next-redux-wrapper`.

Looking forward to hearing any feedback.

**Comment by phryneas** [comment]
Why do you want to avoid that package? It's essentially impossible to use a Redux store with SSR without it.

**Reply by cam-eo** [reply]
I understand it does not work without it and that's why it seems like a common enough problem that maybe it can be covered IN RTK. I'm not trying to complain I'm just wondering if there is a plan to include it in RTK.

**Reply by markerikson** [reply]
It's kind of like with `redux-persist`:

- It's too specific a use case to include in RTK
- The other library already exists and is reasonably widely used, so there's no reason for us to reinvent the wheel here

**Reply by phryneas** [reply]
It is very explicit to SSR scenarios. If you do not use `next-redux-wrapper`, all your SSR renders will use the same store - you will mix data of different users with each other.

Redux and Redux Toolkit are (mostly) independent of the framework used. You can use them with React, Lit, Angular, Svelte, Vue, whatever. These packages are the wrong place to add support for one specific Framework for React (I mean, even for React, there are also Redwood, Gatsby, Remix and a dozen others we would need to start supporting if we started with one).

`next-redux-wrapper` exists. Why should we reimplement it?

**Reply by cam-eo** [reply]
True I guess it's okay...but confidence in these packages for long-term use is kind of a risk for projects making a decision on state management. 

I know some teams that have opted to build their own script to avoid 3rd party dependency for long term confidence.

Also, the fact that packages like these are widely used, doesn't that make my point that it seems to be common enough to cover?

Again, not complaining just wondering. I guess the fact that RTK mentions `next-redux-wrapper` in the documentation gives some confidence. But if its notable in docs then perhaps important to have included

IDK just wondering is all...

**Reply by phryneas** [reply]
> True I guess it's okay...but confidence in these packages for long-term use is kind of a risk for projects making a decision on state management.

Seriously? It's open source. It saves you a ton of work. If it ever stops being maintained, you can fork it and take over, still saving you tons of work.

> Also, the fact that packages like these are widely used, doesn't that make my point that it seems to be common enough to cover?

We are like 3 people caring for all of this in our free time. Do you think we get paid for this? I personally get about 50$ a month in Github Sponsors, it's not different for the others.  

How much more workload do you want to add because it makes sense to you that we do it, while there is already a person that also sacrifices their spare time to already maintain a perfectly viable tool for this exact purpose? That's not how Open Source works.

Also: To my knowledge none of us uses Next for SSR. If you want a tool by an expert in the field and not amateurs, you better go with the established tool instead of asking us to write one.

**Reply by cam-eo** [reply]
Perhaps for SSR something like Tan Stack Query and a light Client Side State Manager like Jotai is a the way to go

@markerikson @phryneas thanks for the insights

**Reply by cam-eo** [reply]
> We are like 3 people caring for all of this in our free time. Do you think we get paid for this? I personally get about 50$ a month in Github Sponsors, it's not different from the others.

> How much more workload do you want to add because it makes sense to you that we do it, while there is already a person that also sacrifices their spare time to already maintain a perfectly viable tool for this exact purpose. That's not how Open Source works.

Wow I really didn't realize how limited resources are. The community is definitely in debt to you folks that work hard to maintain the package.

Knowing that resources are slim to none...totally understand.

**Reply by phryneas** [reply]
You will face the exact same problems with Jotai. SSR is an additional level of complexity. Nobody who maintains a state management library will have the stamina to support two dozen different SSR methods. If you are lucky you will find a third party library adding support for that.

To my knowledge nothing in TanStack Query is optimized for SSR with Next. It just gives you hints in the docs on how to create a new `QueryClient` for each Request. Essentially how to do the job that next-redux-wrapper does for you - but by hand.

Both of these tools are also maintained by 1-2 people, mostly next to normal work (Tanner is working on that stuff full-time, but on many more projects). Resources are so slim everywhere in the Open Source ecosystem.

**Reply by cam-eo** [reply]
Makes a lot of sense....I've been a bit disconnected from open source contribution world. Really makes me feel like I should contribute more.

When I mentioned Jotai, I mean only for Client Side state management for dynamic stuff like animations and Tanner only for SSR/Query stuff.

But I appreciate all the honesty🙏🏾

**Reply by phryneas** [reply]
No matter at which of your dependencies you look, you have a 90% chance it looks like this ;)

[<img src="https://imgs.xkcd.com/comics/dependency_2x.png" />](https://xkcd.com/2347/)

I can only recommend you don't get hung up on your distrust on projects like `next-redux-wrapper`. That will be a rabbit hole you don't want to get into ;)
